Fix for PR #9468: Added ShellyWiFi interface to address Type issue for properties

jactmp opened the pull request:
Proposed fixes for #9468 (PR on top of #9468’s changes).

Fixed compilation error  "TS2322: Type 'string' is not assignable to type '{ ID: number; type: number; }'."
- added `nterface ShellyWiFi`
- added `<"shellyWiFiSetupCluster", ShellyWiFi>` to custom entities.

Fell free to reject my input. It's first PR ever.
~

Koenkk commented on src/devices/shelly.ts:
```diff
@@ -1,34 +1,247 @@
+                cluster: "shellyWiFiSetupCluster",
+                access: "STATE_GET",
+                description: "IP of the device",
+                name: "IP",
```
```suggestion
                name: "ip",
```

jactmp commented on src/devices/shelly.ts:
```diff
@@ -1,34 +1,247 @@
+                cluster: "shellyWiFiSetupCluster",
+                access: "STATE_GET",
+                description: "Wifi Status",
+                name: "Wifi_Status",
```
Sure. Some suggestion needed though. 
Shelly originally uses such naming for attributes.
```
....
7	StaticIP	character string	read/write
8	NetMask	    character string	read/write
9	Gateway	    character string	read/write
10	NameServer	character string	read/write
```
Those name where used as descriptions and attribute's name (lowercased). Cool. Now, the question is what's the proper name for:
7/ `static_ip` or `staticip`?
8/ `net_mask` or `netmask`?
10/ `name_server` or `nameserver`?

Koenkk replied:
I agree with your personal preferences

jactmp commented:
Seems like, contrary to what i remembered from ifconfig output) use of net_mask, and name_server with underline inside is quite common.  So I updated attributes to snake_case (including net_mask, name_server) in devices/shelly.ts as requested (commit 79d5a699e) + resolved divergence with a merge commit (ec3855b). Ready for review.

jactmp commented:
I've just noticed a problem. "action" name cannot be used as its sort of reserved name in z2m? I see in GUI:
_Action
Triggered action (e.g. a button click)
N/A_
instead of Apply/Reset buttons. 
I replaced "action" with "action_" and problem is fixed. Of course final name should be different - something more clean and close to original name.

jactmp commented:
Ok, I noticed more problems. Attributes read/written using GUI or mqtt are retained and published through MQTT any time some state changes. For most attributes it's OK, but not always. 

Problems:
- `password`: this fileld should NEVER be published to MQTT in my opinion - because, well, it's password, and secondly it's part of device state that is write only according to Shelly's specs,
- `action`: similar story. It's write-only "one-time" action, not state, and should not be pusblished to mqtt or at least should not be cached in the system (sort of one only publish). 

For now I have no idea how to address such problems. Is it even possible to make z2m not sending certain attributes to mqtt and/or not retain it's states? 
I thing feature should not be merged to stable branch until problems above are solved.

jactmp commented:
I don't get it. For definition
```
            m.text<"shellyWiFiSetupCluster", ShellyWiFi>({
                cluster: "shellyWiFiSetupCluster",
                access: "SET",
                description: "Wifi Password",
                name: "password",
                attribute: "password",
                entityCategory: "config",
                validate(value: unknown) {
                    assertString(value);
                },
                zigbeeCommandOptions: {
                    profileId: ZSpec.CUSTOM_SHELLY_PROFILE_ID,
                },
            }),
```
where `access` is `SET` (not `STATE_SET`), as soon as password is set, it is cached and send in every MQTT concerning Shelly device afterwards. No way how to address it and why it is as it is. Seems like this is  blocking point for merging WiFi feature to master branch? I can push updated PR with snake_case names ofc (I changed name for Action attribute to `actions` to address reserved name problem).

Koenkk commented:
> I've just noticed a problem. "action" name cannot be used as its sort of reserved name in z2m? I see in GUI: _Action Triggered action (e.g. a button click) N/A_ instead of Apply/Reset buttons. I replaced "action" with "action_" and problem is fixed. Of course final name should be different - something more clean and close to original name.

`action` is indeed reserved for things like button presses, please use a different name.


> password: this fileld should NEVER be published to MQTT in my opinion - because, well, it's password, and secondly it's part of device state that is write only according to Shelly's specs,

Can you try changing this line:

https://github.com/Koenkk/zigbee-herdsman-converters/blob/master/src/lib/modernExtend.ts#L2684

to:

```js
if (access == ea.SET) {
return {state: {[key]: value}};
}

jactmp commented:
I implemented your proposal but modified (not returning anything results in error - no response). I added
```
if (access === ea.SET) {
                          return {state: {[key]: undefined}};
} else {
                          return {state: {[key]: value}};
}
```
to 
`export function text<...` (password attribute )
and
`export function numeric<...` (action attribute)

Some initial tests after this modification. 
- password is not published in mqtt,
- password disappears from GUI after refresh 

Looks good.

jactmp commented:
Ok. Seems like I got sort of random error yesterday. I made some tests today again. I cleared cache (deleted `state.json`) every time before starting z2m. I added 
```ts
                          if (access !== ea.SET) {
                              return {state: {[key]: value}};
```
to `text` function. It works as expected - no password in MQTT nor in state.json (after stoping z2m). 
Actually i added conditional return to `enumLookup` function as well. Otherwise `"actions" : "Apply"` is published in MQTT message, which makes no sense it seems, as it's SET (one time write-only) attribute as well, not permanent "current-device-state" information.

So, what's next steps? I noticed another small issue, but it could be addressed later (not critical).

shaysudm commented:
@jactmp 
can you please also finish the implementation for the RPC cluster?

Koenkk commented:
> 3/ actions and password are SET only, but still cached and published in MQTT in every status message.
You provided solution. It requires changes to core functionality modernExtend. At least to text and enumLookup, probably to all other types for consistency? Challenging stuff as it may affect device handlers that use this library. Definitely not task for me - junior contributor.

Please include the changes in this PR

jactmp commented:
GUI (default) looks now like below:
![wifi](https://github.com/user-attachments/assets/b797f4cc-6fbc-46a7-b44c-05037f0e14db)
